### PR TITLE
AEA-1335 Feature/go lib aealite

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR introduces a lightweight AEA library for Golang. It borrows code from p2p_libp2p

## Fixes

-

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: libs/go/aealite/agent_test.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+	if agent.Identity.PrivateKey != private_key {
+		t.Fatal("AgentIdentity.PrivateKey not set")
+	}
+
+	if !agent.connection.Initialised() {
+		t.Fatal("connection not initialised")
+	}
+
+	// err = agent.Start()
```
Comment (DavidMinarsch) [inline libs/go/aealite/agent_test.go:84]: @lrahmani try enable this and see the registration failure. What is going wrong?

File: libs/go/aealite/connections/p2pclient.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	client.out_queue = make(chan *protocols.Envelope, 10)
+	go client.listen_for_envelopes()
+	logger.Info().Msg("connected to p2p node")
+
+	client.connected = true
+
+	return nil
+}
+
+func (client *P2PClientApi) register() error {
```
Comment (DavidMinarsch) [inline libs/go/aealite/connections/p2pclient.go:163]: This method is likely not quite correct @lrahmani

File: libs/go/aealite/agent.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+const (
+	DefaultLedger = "fetchai"
+)
+
+type Agent struct {
+	Wallet     *wallet.Wallet
+	connection connections.Connection
+}
+
+func (agent *Agent) InitFromEnv() error {
```
Comment (dilwariamukul) [inline libs/go/aealite/agent.go:40]: Can we have interface of Agent functions , in future if we have more types of Agents , and we come across a different implementations , then it wont be structured anymore, but if we dont have more type of Agent types then no need of interface of all the functions.

File: libs/go/aealite/connections/tcpsocket.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	_, err := sock.conn.Write(buf)
+	logger.Debug().Msgf("wrote data to pipe: %d bytes", size)
+	return err
+}
+
+func (sock *TCPSocketChannel) Disconnect() error {
+	return sock.conn.Close()
+}
+
+func NewSocket(address string, port uint16) Socket {
```
Comment (dilwariamukul) [inline libs/go/aealite/connections/tcpsocket.go:73]: From where we are triggering this ?

File: libs/go/aealite/wallet/utils.go [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+		"ethereum": VerifyEthereumSignatureETH,
+	}
+)
+
+// AddressFromPublicKey get wallet address from public key associated with ledgerId
+func AddressFromPublicKey(ledgerId string, publicKey string) (string, error) {
+	if addressFromPublicKey, found := addressFromPublicKeyTable[ledgerId]; found {
+		return addressFromPublicKey(publicKey)
+	}
+	return "", errors.New("Unsupported ledger " + ledgerId + "for AddressFromPublicKey")
```
Comment (dilwariamukul) [inline libs/go/aealite/wallet/utils.go:61]: we can have errors utility to call common errors , its fine for now

File: libs/go/aealite/wallet/utils.go [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+func BTCPubKeyFromEthereumPublicKey(publicKey string) (*btcec.PublicKey, error) {
+	return BTCPubKeyFromUncompressedHex(publicKey[2:])
+}
+
+// ConvertStrEncodedSignatureToDER
+// References:
+//  - https://github.com/fetchai/agents-aea/blob/main/aea/crypto/cosmos.py#L258
+//  - https://github.com/btcsuite/btcd/blob/master/btcec/signature.go#L47
+func ConvertStrEncodedSignatureToDER(signature []byte) []byte {
+	rb := signature[:len(signature)/2]
```
Comment (dilwariamukul) [inline libs/go/aealite/wallet/utils.go:108]: first check signature is not nil

File: libs/go/aealite/wallet/utils.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+//  - https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L452-L459
+//  - https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L404
+func RecoverAddressFromEthereumSignature(message []byte, signature string) (string, error) {
+	// prepare signature
+	sigBytes, err := hexutil.Decode(signature)
+	if err != nil {
+		return "", err
+	}
+
+	if sigBytes[64] != 27 && sigBytes[64] != 28 {
```
Comment (dilwariamukul) [inline libs/go/aealite/wallet/utils.go:239]: keep 27 and 28 in constants , i dont the use of these numbers , but if other numbers come into picture or need to change them

File: libs/go/aealite/agent.go (hunk 3 above)
Comment (DavidMinarsch) [reply libs/go/aealite/agent.go:40]:
```
type Agent interface {
	InitFromEnv() error
        ...
}
```

File: libs/go/aealite/connections/tcpsocket.go (hunk 4 above)
Comment (DavidMinarsch) [reply libs/go/aealite/connections/tcpsocket.go:73]: line 125 in p2pclient.go

**Pull request merged**